RPC: Reject RPC requests with same named parameter specified multiple times

ryanofsky opened the pull request:
Make the JSON-RPC server reject requests with the same named parameter specified multiple times, instead of silently overwriting earlier parameter values with later ones.

Generally JSON keys are supposed to unique, and their order isn't supposed to be significant, so having the server silently discard duplicate keys is error-prone. Most likely if an RPC client is sending a request with duplicate keys it means something is wrong with the request and there should be an error.

After this change, named parameters are still allowed to specified multiple times on the `bitcoin-cli` command line, since `bitcoin-cli` automatically replaces earlier values with later values before sending the JSON-RPC request. This makes sense, since it's not unusual for the order of command line options to be significant or for later command line options to override earlier ones.

ryanofsky committed:
test: Add RPC tests for same named parameter specified more than once

Current behavior isn't ideal and will be changed in upcoming commits, but it's
useful to have test coverage regardless.

MarcoFalke reported the case of bitcoin-cli positional arguments overwriting
the named "args" parameter in
https://github.com/bitcoin/bitcoin/pull/19762#discussion_r1035761471

ryanofsky committed:
rpc: Make it an error server-side to specify same named parameter multiple times

Specifying same named parameter multiple times is still allowed by bitcoin-cli.
The client implementation overwrites earlier option values with later ones
before sending to server. This is tested by interface_bitcoin_cli.py

Rationale for allowing client parameters to be specified multiple times in
bitcoin-cli is that this behavior has been supported for a long time, and that
when using the command line interactively, it can be convenient to override
earlier option values with new values without having to go back and remove the
old value.

But for the RPC server, there isn't really a good use-case for earlier values
to be discarded if multiple values are specified. JSON keys are generally
supposed to be unique and if they aren't it's probably an indication of some
problem generating the RPC request.

ryanofsky committed:
bitcoin-cli: Make it an error to specify the "args" parameter two different ways

MarcoFalke reported the case of positional arguments silently overwriting the
named "args" parameter in bitcoin-cli
https://github.com/bitcoin/bitcoin/pull/19762#discussion_r1035761471 and this
behavior is confusing and was not intended when support for "args" parameters
was added to bitcoin-cli in #19762.

Instead of letting one "args" value overwrite the other in the client, just
pass the values to the server verbatim, and let the error be handled server
side.

MarcoFalke commented on src/test/rpc_tests.cpp:
```diff
@@ -89,6 +89,10 @@ BOOST_AUTO_TEST_CASE(rpc_namedparams)
+
     // Make sure named and positional arguments can be combined.
     BOOST_CHECK_EQUAL(TransformParams(JSON(R"({"arg5": 5, "args": [1, 2], "arg4": 4})"), arg_names).write(), "[1,2,null,4,5]");
 
```
nit: Could fix https://github.com/bitcoin/bitcoin/pull/19762#discussion_r1035955247 here as well?

MarcoFalke commented on doc/release-notes-xxxxx.md:
```diff
@@ -0,0 +1,4 @@
+JSON-RPC
+---
+
+The JSON-RPC server now rejects requests where same named parameter is specified multiple times, instead of silently overwriting earlier parameter values with later ones.
```
```suggestion
The JSON-RPC server now rejects requests where same named parameter is specified multiple times, instead of silently overwriting earlier parameter values with later ones. (#26628)
```

stickies-v replied:
File name still has to be updated too

maflcko reviewed (approved):
lgtm, left two nits

stickies-v reviewed (commented):
Concept ACK

stickies-v commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -90,6 +90,10 @@ def run_test(self):
 
+        self.log.info("Test that later cli named arguments values silently overwrite earlier ones")
+        assert_equal(self.nodes[0].cli("-named", "echo", "arg0=0", "arg1=1", "arg2=2", "arg1=3").send_cli(), ['0', '3', '2'])
+        assert_raises_rpc_error(-8, "Parameter args specified multiple times", self.nodes[0].cli("-named", "echo", "args=[0,1,2,3]", "4", "5", "6", ).send_cli)
```
I'm not sure if it's worth changing, but seeing this test made me realize this error message could be confusing feedback to a user? They're not specifying `args` multiple times _explicitly_ - the fact that it's labelled `args` internally may not be immediately obvious? But also maybe it's a bit unnecessary to add special error handling for the `args` case.

stickies-v commented on src/rpc/server.cpp:
```diff
@@ -399,7 +399,10 @@ static inline JSONRPCRequest transformNamedArguments(const JSONRPCRequest& in, c
     std::unordered_map<std::string, const UniValue*> argsIn;
     for (size_t i=0; i<keys.size(); ++i) {
-        argsIn[keys[i]] = &values[i];
+        auto inserted = argsIn.emplace(keys[i], &values[i]);
+        if (!inserted.second) {
```
nit: I think it's reasonable enough to skip using the intermediate `inserted` variable entirely, but if you think it's better for readability I'd suggest going all the way and make `inserted` the actual boolean directly?
```suggestion
        auto [_, inserted] = argsIn.emplace(keys[i], &values[i]);
        if (!inserted) {
```

stickies-v reviewed (commented):
Approach ACK 763d3947d34fa71f1a0816c95fb6e206507e2ece

luke-jr commented:
Multiple keys with the same name is IIRC a supported feature of JSON, but concept ACK on rejecting them rather than ignoring all but one.

I think `bitcoin-cli` should also reject duplicate parameters, though. It's not obvious to a user that later ones will override earlier ones - they might expect to be able to specify some parameters twice and get the behaviour of both somehow. For example, `getblock blockhash=... blockhash=...` would logically make more sense as a batch call (though I don't think it's practical to support that, so an error is best IMO).

promag commented on src/rpc/client.cpp:
```diff
@@ -299,7 +299,7 @@ UniValue RPCConvertNamedValues(const std::string &strMethod, const std::vector<s
 
     if (!positional_args.empty()) {
-        params.pushKV("args", positional_args);
+        params.__pushKV("args", positional_args);
```
763d3947d34fa71f1a0816c95fb6e206507e2ece

Using` args=` is fine but not when used along with positional arguments. Agree with @stickies-v, it's not that clear from the error message what the user should do.

An alternative is to fail here if `params` already contain `args`.

promag reviewed (approved):
Code review ACK 763d3947d34fa71f1a0816c95fb6e206507e2ece.

promag commented:
@luke-jr suggestion to fail on duplicate named parameters also makes sense to me, either implemented here or in a different branch.

ryanofsky committed:
test: Suggested cleanups for rpc_namedparams test

No changes in behavior, just implements review suggestions from

https://github.com/bitcoin/bitcoin/pull/19762#discussion_r1025573943
https://github.com/bitcoin/bitcoin/pull/19762#discussion_r1035955247
https://github.com/bitcoin/bitcoin/pull/26628#discussion_r1038765894

ryanofsky reviewed (commented):
Thanks for the reviews!

Updated 763d3947d34fa71f1a0816c95fb6e206507e2ece -> 8c3ff7d52ae3314959e1e66da8718a3f0d30abaa ([`pr/nmult.1`](https://github.com/ryanofsky/bitcoin/commits/pr/nmult.1) -> [`pr/nmult.2`](https://github.com/ryanofsky/bitcoin/commits/pr/nmult.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/nmult.1..pr/nmult.2)) with suggestions.

re: https://github.com/bitcoin/bitcoin/pull/26628#issuecomment-1340616084

> I think `bitcoin-cli` should also reject duplicate parameters, though. It's not obvious to a user that later ones will override earlier ones - they might expect to be able to specify some parameters twice and get the behaviour of both somehow. For example, `getblock blockhash=... blockhash=...` would logically make more sense as a batch call (though I don't think it's practical to support that, so an error is best IMO).

I think it's pretty normal for command line tools to let later option values override earlier ones. Our binaries `bitcoind`, `bitcoin-qt`, `bitcoin-wallet` and `bitcoin-cli` all support this for options which don't accept multiple values. In general I think it's good for command line tools to support shortcuts that make scripting easier and remove the need for unnecessary typing and editing, but I can understand the opposite point of view. In any case, i think it would be better to follow up on this in another PR. This PR is trying to avoid the need for a decision by keeping the current behavior unchanged.

maflcko reviewed (commented):
review ACK 8c3ff7d52ae3314959e1e66da8718a3f0d30abaa 🗂

<details><summary>Show signature</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

review ACK 8c3ff7d52ae3314959e1e66da8718a3f0d30abaa 🗂
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUjCxwwAvbXzoDqw5hQHSmtV1qviWaF9zdsfXYeGfXy4TlIyT5jk/pSAmpMj6vqc
gry65aiuJiZ1mP3ywqMXw7+Fs6m/8wJCXl1cBHt43HDLyD1GDMJusVD3A5FsrtPv
1zfZ4zt24cuDnEjImYdf+aWmszO1/PWeXs9n7WJ9AxhV7lE/740lHgjDj8or/yLz
KwuHwmGE1Z1pwV8t9aM0VLgPB6VWcAfXREoxeRbftF84UZ2/tTwER3YNw9Oh2Mbs
hOh1dCDUyzizziKSWlpzXUjPBLeGoe70W0W2CP9iukl/sRpLgEcDooPwqI9DQ1wv
htJqNJrXN/SDtnSulGSNC57500g00J/Jy/Xp+flridDtdZpqQB9LPx0KE+U+0oYz
vRaF9DsrOWtR6QSLqheFwjusosOKm4g7OV8DRrCIdbZVwiwZDeNFW/bfEBrTRvvk
T+efue7d+zc8Gq/oWQ+hRMUfj0GkESfWDlUZok/SqWvKWQnA82bvMi/qmb7BhPB9
swtWkjI9
=ECGn
-----END PGP SIGNATURE-----
```


</details>

kristapsk reviewed (approved):
ACK 8c3ff7d52ae3314959e1e66da8718a3f0d30abaa

stickies-v reviewed (approved):
ACK 8c3ff7d52

No strong view on the `bitcoin-cli` behaviour. Intuitively I'd have implemented it to fail on duplicated names, but I think both approaches have merit. Happy to start with the current behaviour and potentially change in the future.

LarryRuane commented:
> I think it's pretty normal for command line tools to let later option values override earlier ones. Our binaries `bitcoind`, `bitcoin-qt`, `bitcoin-wallet` and `bitcoin-cli` all support this for options which don't accept multiple values. In general I think it's good for command line tools to support shortcuts that make scripting easier and remove the need for unnecessary typing and editing, but I can understand the opposite point of view.

FWIW, I also favor the behavior of later arguments overriding earlier ones (when multiple don't make sense). I'm not sure if the following is what you're referring to, but I think it's helpful to be able to change a default in a shell script or an alias, but also allow the user to override the new default, something like:
```
alias printblock='bitcoin-cli -named getblock verbose=true'
```
Then you can do: `printblock -blockhash=00...9` to get verbose output, or `printblock -verbose=false -blockhash=00...9` to get a non-verbose result.
